server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -325,7 +327,8 @@ public IndexShard(
         final Runnable globalCheckpointSyncer,
         final RetentionLeaseSyncer retentionLeaseSyncer,
         final CircuitBreakerService circuitBreakerService,
-        @Nullable final SegmentReplicationCheckpointPublisher checkpointPublisher
+        @Nullable final SegmentReplicationCheckpointPublisher checkpointPublisher,
```

Comment:
I think we'll need to ensure these refresh listeners run in sequence and perhaps combine them when both features are enabled. I'd imagine the segrep checkpoint publisher needs to wait until files have completed upload to a remote store before alerting replicas new segments are available to pull from that store.  The alternative is implementing a pull strategy where replicas periodically ping the store for metadata on its latest checkpoint.  With node to node replication we are using push through this segrep listener.

Reply:
As I understand, listeners are called in sequence. If not, as you suggested, we need a mechanism to pass the info on when the upload is complete. I am not sure about combining these listeners though. It will just make them tightly coupled.

Reply:
I'm not sure there is much risk in creating one listener here given how much these features overlap.  Or extending `RemoteStoreRefreshListener `to take a `SegmentReplicationCheckpointPublisher` and publishing within the remote store listener only if the segrep index setting is enabled.  We'd then disable the `checkpointListener` if remote store & segrep are both enabled.  Hiding all of that logic in a single listener makes sense to me.